File: packages/ash-node/src/crypto-properties.test.ts
/**
 * Cryptographic Properties Comprehensive Tests
 *
 * Tests for cryptographic properties covering:
 * - Avalanche effect (small input changes cause large output changes)
 * - Collision resistance
 * - Timing-safe comparison
 * - Entropy analysis
 * - Hash distribution
 * - Key derivation properties
 */

import { describe, it, expect, beforeAll } from 'vitest';
import * as crypto from 'crypto';
import {
  ashInit,
  ashHashBody,
  ashHashProof,
  ashHashScope,
  ashDeriveClientSecret,
  ashBuildProof,
  ashGenerateNonce,
  ashGenerateContextId,
  ashGenerateContextId256,
  ashTimingSafeEqual,
  ashNormalizeBinding,
  MIN_NONCE_BYTES,
  MIN_NONCE_HEX_CHARS,
  MAX_NONCE_LENGTH,
} from './index';

// Initialize WASM
beforeAll(() => {
  try {
    ashInit();
  } catch {
    // WASM not available, tests will use native fallback
  }
});

describe('Cryptographic Properties Comprehensive Tests', () => {
  describe('Avalanche Effect - Hashing', () => {
    it('single bit change produces vastly different hash', () => {
      const input1 = 'Hello, World!';
      const input2 = 'Hello, World?'; // Changed ! to ?

      const hash1 = ashHashBody(input1);
      const hash2 = ashHashBody(input2);

      // Count differing hex characters
      let diffCount = 0;
      for (let i = 0; i < hash1.length; i++) {
        if (hash1[i] !== hash2[i]) diffCount++;
      }

      // Should differ in roughly 50% of characters (avalanche effect)
      expect(diffCount).toBeGreaterThan(hash1.length * 0.3);
    });

    it('adjacent values produce uncorrelated hashes', () => {
      const hashes: string[] = [];
      for (let i = 0; i < 100; i++) {
        hashes.push(ashHashBody(i.toString()));
      }

      // Check that consecutive hashes are significantly different
      for (let i = 0; i < hashes.length - 1; i++) {
        let diffCount = 0;
        for (let j = 0; j < hashes[i].length; j++) {
          if (hashes[i][j] !== hashes[i + 1][j]) diffCount++;
        }
        // Each pair should differ significantly
        expect(diffCount).toBeGreaterThan(10);
      }
    });

    it('prefix addition changes entire hash', () => {
      const base = 'test message';
      const prefixed = 'a' + base;

      const hash1 = ashHashBody(base);
      const hash2 = ashHashBody(prefixed);

      let diffCount = 0;
      for (let i = 0; i < hash1.length; i++) {
        if (hash1[i] !== hash2[i]) diffCount++;
      }

      expect(diffCount).toBeGreaterThan(hash1.length * 0.3);
    });

    it('suffix addition changes entire hash', () => {
      const base = 'test message';
      const suffixed = base + 'z';

      const hash1 = ashHashBody(base);
      const hash2 = ashHashBody(suffixed);

      let diffCount = 0;
      for (let i = 0; i < hash1.length; i++) {
        if (hash1[i] !== hash2[i]) diffCount++;
      }

      expect(diffCount).toBeGreaterThan(hash1.length * 0.3);
    });

    it('case change produces different hash', () => {
      const hash1 = ashHashBody('HELLO');
      const hash2 = ashHashBody('hello');

      expect(hash1).not.toBe(hash2);

      let diffCount = 0;
      for (let i = 0; i < hash1.length; i++) {
        if (hash1[i] !== hash2[i]) diffCount++;
      }

      expect(diffCount).toBeGreaterThan(hash1.length * 0.3);
    });
  });

  describe('Avalanche Effect - HMAC Proofs', () => {
    it('timestamp change produces vastly different proof', () => {
      const secret = 'secret'.repeat(10);
      const binding = 'GET|/api/test|';
      const bodyHash = ashHashBody('{}');

      const proof1 = ashBuildProof(secret, '1000000000', binding, bodyHash);
      const proof2 = ashBuildProof(secret, '1000000001', binding, bodyHash);

      let diffCount = 0;
      for (let i = 0; i < proof1.length; i++) {
        if (proof1[i] !== proof2[i]) diffCount++;
      }

      expect(diffCount).toBeGreaterThan(proof1.length * 0.3);
    });

    it('secret change produces vastly different proof', () => {
      const binding = 'GET|/api/test|';
      const bodyHash = ashHashBody('{}');
      const timestamp = '1000000000';

      const proof1 = ashBuildProof('secret1', timestamp, binding, bodyHash);
      const proof2 = ashBuildProof('secret2', timestamp, binding, bodyHash);

      let diffCount = 0;
      for (let i = 0; i < proof1.length; i++) {
        if (proof1[i] !== proof2[i]) diffCount++;
      }

      expect(diffCount).toBeGreaterThan(proof1.length * 0.3);
    });

    it('binding change produces vastly different proof', () => {
      const secret = 'secret'.repeat(10);
      const bodyHash = ashHashBody('{}');
      const timestamp = '1000000000';

      const proof1 = ashBuildProof(secret, timestamp, 'GET|/api/test|', bodyHash);
      const proof2 = ashBuildProof(secret, timestamp, 'POST|/api/test|', bodyHash);

      let diffCount = 0;
      for (let i = 0; i < proof1.length; i++) {
        if (proof1[i] !== proof2[i]) diffCount++;
      }

      expect(diffCount).toBeGreaterThan(proof1.length * 0.3);
    });
  });

  describe('Collision Resistance', () => {
    it('no collisions in 10000 random hashes', () => {
      const hashes = new Set<string>();
      for (let i = 0; i < 10000; i++) {
        const input = crypto.randomBytes(32).toString('hex');
        const hash = ashHashBody(input);
        expect(hashes.has(hash)).toBe(false);
        hashes.add(hash);
      }
      expect(hashes.size).toBe(10000);
    });

    it('no collisions in sequential integers', () => {
      const hashes = new Set<string>();
      for (let i = 0; i < 10000; i++) {
        const hash = ashHashBody(i.toString());
        expect(hashes.has(hash)).toBe(false);
        hashes.add(hash);
      }
      expect(hashes.size).toBe(10000);
    });

    it('no collisions in similar strings', () => {
      const hashes = new Set<string>();
      const base = 'transaction_';
      for (let i = 0; i < 1000; i++) {
        const hash = ashHashBody(base + i);
        expect(hashes.has(hash)).toBe(false);
        hashes.add(hash);
      }
      expect(hashes.size).toBe(1000);
    });

    it('no collisions in proofs with same secret different timestamps', () => {
      const secret = 'testsecret123';
      const binding = 'POST|/api/test|';
      const bodyHash = ashHashBody('{}');

      const proofs = new Set<string>();
      for (let i = 0; i < 1000; i++) {
        const timestamp = (1700000000 + i).toString();
        const proof = ashBuildProof(secret, timestamp, binding, bodyHash);
        expect(proofs.has(proof)).toBe(false);
        proofs.add(proof);
      }
      expect(proofs.size).toBe(1000);
    });
  });

  describe('Timing-Safe Comparison', () => {
    it('equal strings return true', () => {
      const str = 'a'.repeat(64);
      expect(ashTimingSafeEqual(str, str)).toBe(true);
    });

    it('different strings return false', () => {
      const str1 = 'a'.repeat(64);
      const str2 = 'b'.repeat(64);
      expect(ashTimingSafeEqual(str1, str2)).toBe(false);
    });

    it('strings differing by one char return false', () => {
      const str1 = 'a'.repeat(64);
      const str2 = 'a'.repeat(63) + 'b';
      expect(ashTimingSafeEqual(str1, str2)).toBe(false);
    });

    it('different length strings return false', () => {
      expect(ashTimingSafeEqual('short', 'longer')).toBe(false);
      expect(ashTimingSafeEqual('longer', 'short')).toBe(false);
    });

    it('empty strings return true', () => {
      expect(ashTimingSafeEqual('', '')).toBe(true);
    });

    it('timing is consistent regardless of difference position', () => {
      const base = 'a'.repeat(64);
      const positions = [0, 16, 32, 48, 63];
      const iterations = 1000;
      const timings: number[] = [];

      for (const pos of positions) {
        const modified = base.substring(0, pos) + 'b' + base.substring(pos + 1);
        const start = process.hrtime.bigint();
        for (let i = 0; i < iterations; i++) {
          ashTimingSafeEqual(base, modified);
        }
        timings.push(Number(process.hrtime.bigint() - start));
      }

      // All timings should be within reasonable ratio
      const maxTime = Math.max(...timings);
      const minTime = Math.min(...timings);
      const ratio = maxTime / minTime;
      expect(ratio).toBeLessThan(5);
    });
  });

  describe('Entropy Analysis - Nonce Generation', () => {
    it('generates minimum required entropy', () => {
      const nonce = ashGenerateNonce();
      // Default is 32 bytes = 64 hex chars
      expect(nonce.length).toBe(64);
    });

    it('generates cryptographically random nonces', () => {
      const nonces = new Set<string>();
      for (let i = 0; i < 10000; i++) {
        nonces.add(ashGenerateNonce());
      }
      expect(nonces.size).toBe(10000);
    });

    it('nonce bytes have uniform distribution', () => {
      const nibbleCounts = new Array(16).fill(0);
      const iterations = 10000;

      for (let i = 0; i < iterations; i++) {
        const nonce = ashGenerateNonce();
        for (const char of nonce) {
          nibbleCounts[parseInt(char, 16)]++;
        }
      }

      // Check for roughly uniform distribution
      const totalNibbles = iterations * 64;
      const expected = totalNibbles / 16;
      const tolerance = expected * 0.1; // 10% tolerance

      for (const count of nibbleCounts) {
        expect(count).toBeGreaterThan(expected - tolerance);
        expect(count).toBeLessThan(expected + tolerance);
      }
    });

    it('generates nonces with specified byte length', () => {
      expect(ashGenerateNonce(16).length).toBe(32); // 16 bytes = 32 hex chars
      expect(ashGenerateNonce(32).length).toBe(64); // 32 bytes = 64 hex chars
      expect(ashGenerateNonce(64).length).toBe(128); // 64 bytes = 128 hex chars
    });

    it('rejects nonce below minimum bytes', () => {
      expect(() => ashGenerateNonce(MIN_NONCE_BYTES - 1)).toThrow(/at least/);
    });

    it('rejects nonce above maximum bytes', () => {
      const maxBytes = MAX_NONCE_LENGTH / 2;
      expect(() => ashGenerateNonce(maxBytes + 1)).toThrow(/cannot exceed/);
    });
  });

  describe('Entropy Analysis - Context ID Generation', () => {
    it('generates unique context IDs', () => {
      const ids = new Set<string>();
      for (let i = 0; i < 10000; i++) {
        ids.add(ashGenerateContextId());
      }
      expect(ids.size).toBe(10000);
    });

    it('context ID has correct format', () => {
      const id = ashGenerateContextId();
      expect(id).toMatch(/^ash_[0-9a-f]{32}$/);
    });

    it('context ID 256-bit version has correct format', () => {
      const id = ashGenerateContextId256();
      expect(id).toMatch(/^ash_[0-9a-f]{64}$/);
    });

    it('256-bit context IDs are unique', () => {
      const ids = new Set<string>();
      for (let i = 0; i < 10000; i++) {
        ids.add(ashGenerateContextId256());
      }
      expect(ids.size).toBe(10000);
    });
  });

  describe('Key Derivation Properties', () => {
    it('derived secret is deterministic', () => {
      const nonce = 'a'.repeat(64);
      const contextId = 'ctx_test';
      const binding = 'POST|/api/test|';

      const secret1 = ashDeriveClientSecret(nonce, contextId, binding);
      const secret2 = ashDeriveClientSecret(nonce, contextId, binding);

      expect(secret1).toBe(secret2);
    });

    it('different nonces produce different secrets', () => {
      const contextId = 'ctx_test';
      const binding = 'POST|/api/test|';

      const secret1 = ashDeriveClientSecret('a'.repeat(64), contextId, binding);
      const secret2 = ashDeriveClientSecret('b'.repeat(64), contextId, binding);

      expect(secret1).not.toBe(secret2);
    });

    it('different context IDs produce different secrets', () => {
      const nonce = 'a'.repeat(64);
      const binding = 'POST|/api/test|';

      const secret1 = ashDeriveClientSecret(nonce, 'ctx_test1', binding);
      const secret2 = ashDeriveClientSecret(nonce, 'ctx_test2', binding);

      expect(secret1).not.toBe(secret2);
    });

    it('different bindings produce different secrets', () => {
      const nonce = 'a'.repeat(64);
      const contextId = 'ctx_test';

      const secret1 = ashDeriveClientSecret(nonce, contextId, 'GET|/path1|');
      const secret2 = ashDeriveClientSecret(nonce, contextId, 'GET|/path2|');

      expect(secret1).not.toBe(secret2);
    });

    it('derived secret has correct length', () => {
      const nonce = 'a'.repeat(64);
      const contextId = 'ctx_test';
      const binding = 'POST|/api/test|';

      const secret = ashDeriveClientSecret(nonce, contextId, binding);
      expect(secret.length).toBe(64); // SHA-256 hex output
    });

    it('derived secret is lowercase hex', () => {
      const nonce = 'A'.repeat(64);
      const contextId = 'ctx_test';
      const binding = 'POST|/api/test|';

      const secret = ashDeriveClientSecret(nonce, contextId, binding);
      expect(secret).toMatch(/^[0-9a-f]{64}$/);
    });
  });

  describe('Hash Distribution', () => {
    it('hash output is evenly distributed across hex chars', () => {
      const charCounts: Record<string, number> = {};
      for (let c = 0; c < 16; c++) {
        charCounts[c.toString(16)] = 0;
      }

      for (let i = 0; i < 1000; i++) {
        const hash = ashHashBody(i.toString());
        for (const char of hash) {
          charCounts[char]++;
        }
      }

      // Check distribution
      const values = Object.values(charCounts);
      const mean = values.reduce((a, b) => a + b, 0) / values.length;
      const variance = values.reduce((sum, v) => sum + Math.pow(v - mean, 2), 0) / values.length;
      const stdDev = Math.sqrt(variance);

      // Standard deviation should be relatively small compared to mean
      expect(stdDev / mean).toBeLessThan(0.1);
    });

    it('hash first char distribution is uniform', () => {
      const firstCharCounts: Record<string, number> = {};
      for (let c = 0; c < 16; c++) {
        firstCharCounts[c.toString(16)] = 0;
      }

      for (let i = 0; i < 10000; i++) {
        const hash = ashHashBody(crypto.randomBytes(32).toString('hex'));
        firstCharCounts[hash[0]]++;
      }

      // Each char should appear roughly 625 times (10000 / 16)
      const values = Object.values(firstCharCounts);
      const min = Math.min(...values);
      const max = Math.max(...values);

      // Max should not be more than 2x min for uniform distribution
      expect(max / min).toBeLessThan(2);
    });
  });

  describe('Proof Properties', () => {
    it('proof is deterministic for same inputs', () => {
      const secret = 'testsecret';
      const timestamp = '1700000000';
      const binding = 'POST|/api/test|';
      const bodyHash = ashHashBody('{}');

      const proof1 = ashBuildProof(secret, timestamp, binding, bodyHash);
      const proof2 = ashBuildProof(secret, timestamp, binding, bodyHash);

      expect(proof1).toBe(proof2);
    });

    it('proof changes with any input change', () => {
      const base = {
        secret: 'testsecret',
        timestamp: '1700000000',
        binding: 'POST|/api/test|',
        bodyHash: ashHashBody('{}'),
      };

      const baseProof = ashBuildProof(base.secret, base.timestamp, base.binding, base.bodyHash);

      // Change secret
      expect(ashBuildProof('different', base.timestamp, base.binding, base.bodyHash)).not.toBe(baseProof);

      // Change timestamp
      expect(ashBuildProof(base.secret, '1700000001', base.binding, base.bodyHash)).not.toBe(baseProof);

      // Change binding
      expect(ashBuildProof(base.secret, base.timestamp, 'GET|/other|', base.bodyHash)).not.toBe(baseProof);

      // Change body hash
      expect(ashBuildProof(base.secret, base.timestamp, base.binding, ashHashBody('different'))).not.toBe(baseProof);
    });

    it('proof is 64 hex characters', () => {
      const proof = ashBuildProof('secret', '1700000000', 'GET|/path|', ashHashBody('{}'));
      expect(proof).toMatch(/^[0-9a-f]{64}$/);
    });
  });

  describe('Scope Hash Properties', () => {
    it('scope hash is deterministic', () => {
      const scope = ['field1', 'field2'];
      const hash1 = ashHashScope(scope);
      const hash2 = ashHashScope(scope);
      expect(hash1).toBe(hash2);
    });

    it('scope hash is order-independent', () => {
      const hash1 = ashHashScope(['a', 'b', 'c']);
      const hash2 = ashHashScope(['c', 'a', 'b']);
      expect(hash1).toBe(hash2);
    });

    it('scope hash handles duplicates', () => {
      const hash1 = ashHashScope(['a', 'b']);
      const hash2 = ashHashScope(['a', 'b', 'a', 'b']);
      expect(hash1).toBe(hash2);
    });

    it('different scopes produce different hashes', () => {
      const hash1 = ashHashScope(['a']);
      const hash2 = ashHashScope(['b']);
      expect(hash1).not.toBe(hash2);
    });

    it('empty scope returns empty string', () => {
      expect(ashHashScope([])).toBe('');
    });
  });

  describe('Chain Hash Properties', () => {
    it('chain hash is deterministic', () => {
      const proof = 'a'.repeat(64);
      const hash1 = ashHashProof(proof);
      const hash2 = ashHashProof(proof);
      expect(hash1).toBe(hash2);
    });

    it('different proofs produce different chain hashes', () => {
      const hash1 = ashHashProof('a'.repeat(64));
      const hash2 = ashHashProof('b'.repeat(64));
      expect(hash1).not.toBe(hash2);
    });

    it('chain hash has avalanche effect', () => {
      const proof1 = 'a'.repeat(64);
      const proof2 = 'a'.repeat(63) + 'b';

      const hash1 = ashHashProof(proof1);
      const hash2 = ashHashProof(proof2);

      let diffCount = 0;
      for (let i = 0; i < hash1.length; i++) {
        if (hash1[i] !== hash2[i]) diffCount++;
      }

      expect(diffCount).toBeGreaterThan(hash1.length * 0.3);
    });
  });
});

File: packages/ash-go/crypto_properties_test.go
package ash

import (
	"fmt"
	"math/rand"
	"strings"
	"testing"
	"time"
)

// ============================================================================
// CRYPTOGRAPHIC PROPERTY TESTS
// ============================================================================

// --- Avalanche Effect Tests ---

func TestAvalancheEffectHashBody(t *testing.T) {
	input := "test input string"
	hash1 := HashBody(input)

	// Single bit change should cause ~50% bit change
	for i := 0; i < len(input); i++ {
		modified := input[:i] + string(input[i]^1) + input[i+1:]
		hash2 := HashBody(modified)

		if hash1 == hash2 {
			t.Errorf("Single byte change at position %d should change hash", i)
		}

		// Count differing characters (simple proxy for bit difference)
		diff := 0
		for j := 0; j < len(hash1); j++ {
			if hash1[j] != hash2[j] {
				diff++
			}
		}
		// Expect significant change (at least 25% of chars)
		if diff < 16 {
			t.Errorf("Avalanche effect weak at position %d: only %d chars differ", i, diff)
		}
	}
}

func TestAvalancheEffectSecret(t *testing.T) {
	nonce := strings.Repeat("a", 32)
	contextID := "test-ctx"
	binding := "POST|/api|"

	secret1 := DeriveClientSecret(nonce, contextID, binding)

	// Modify nonce
	modifiedNonce := "b" + nonce[1:]
	secret2 := DeriveClientSecret(modifiedNonce, contextID, binding)

	if secret1 == secret2 {
		t.Error("Different nonce should produce different secret")
	}

	// Count differences
	diff := 0
	for i := 0; i < len(secret1); i++ {
		if secret1[i] != secret2[i] {
			diff++
		}
	}
	if diff < 16 {
		t.Errorf("Avalanche effect weak: only %d chars differ", diff)
	}
}

func TestAvalancheEffectProof(t *testing.T) {
	secret := strings.Repeat("a", 64)
	timestamp := "1234567890"
	binding := "POST|/api|"
	bodyHash := HashBody(`{}`)

	proof1 := BuildProofV21(secret, timestamp, binding, bodyHash)

	// Modify timestamp by 1
	proof2 := BuildProofV21(secret, "1234567891", binding, bodyHash)

	if proof1 == proof2 {
		t.Error("Different timestamp should produce different proof")
	}

	diff := 0
	for i := 0; i < len(proof1); i++ {
		if proof1[i] != proof2[i] {
			diff++
		}
	}
	if diff < 16 {
		t.Errorf("Avalanche effect weak: only %d chars differ", diff)
	}
}

// --- Collision Resistance Tests ---

func TestCollisionResistanceHash(t *testing.T) {
	hashes := make(map[string]string)
	iterations := 10000

	for i := 0; i < iterations; i++ {
		input := fmt.Sprintf("input_%d_%d", i, time.Now().UnixNano())
		hash := HashBody(input)

		if existing, found := hashes[hash]; found {
			t.Errorf("Collision found: %s and %s both hash to %s", input, existing, hash)
		}
		hashes[hash] = input
	}
}

func TestCollisionResistanceSecret(t *testing.T) {
	secrets := make(map[string]string)
	rng := rand.New(rand.NewSource(time.Now().UnixNano()))
	binding := "POST|/api|"

	for i := 0; i < 10000; i++ {
		nonce := fmt.Sprintf("%032x", rng.Int63())
		contextID := fmt.Sprintf("ctx_%d", i)
		secret := DeriveClientSecret(nonce, contextID, binding)

		key := nonce + "|" + contextID
		if existing, found := secrets[secret]; found {
			t.Errorf("Collision: %s and %s produce same secret", key, existing)
		}
		secrets[secret] = key
	}
}

func TestCollisionResistanceProof(t *testing.T) {
	proofs := make(map[string]string)
	secret := strings.Repeat("a", 64)
	binding := "POST|/api|"

	for i := 0; i < 10000; i++ {
		timestamp := fmt.Sprintf("%d", 1000000000+i)
		bodyHash := HashBody(fmt.Sprintf(`{"i":%d}`, i))
		proof := BuildProofV21(secret, timestamp, binding, bodyHash)

		key := timestamp + "|" + bodyHash
		if existing, found := proofs[proof]; found {
			t.Errorf("Collision: %s and %s produce same proof", key, existing)
		}
		proofs[proof] = key
	}
}

// --- Unpredictability Tests ---

func TestUnpredictabilityHash(t *testing.T) {
	// Sequential inputs should not produce sequential or predictable hashes
	var prevHash string
	for i := 0; i < 1000; i++ {
		hash := HashBody(fmt.Sprintf("%d", i))

		if prevHash != "" {
			// Check that hash doesn't just increment
			if hash == prevHash {
				t.Error("Sequential inputs produced same hash")
			}
		}
		prevHash = hash
	}
}

func TestUnpredictabilitySecret(t *testing.T) {
	contextID := "test-ctx"
	binding := "POST|/api|"
	var prevSecret string

	for i := 0; i < 1000; i++ {
		nonce := fmt.Sprintf("%032d", i)
		secret := DeriveClientSecret(nonce, contextID, binding)

		if prevSecret != "" && secret == prevSecret {
			t.Error("Sequential nonces produced same secret")
		}
		prevSecret = secret
	}
}

// --- Distribution Tests ---

func TestHashDistribution(t *testing.T) {
	// Test that hash output is uniformly distributed
	charCounts := make(map[rune]int)
	iterations := 10000

	for i := 0; i < iterations; i++ {
		hash := HashBody(fmt.Sprintf("input_%d", i))
		for _, c := range hash {
			charCounts[c]++
		}
	}

	// Each hex char (0-9, a-f) should appear roughly equally
	totalChars := iterations * 64
	expectedPerChar := totalChars / 16
	tolerance := float64(expectedPerChar) * 0.2 // 20% tolerance

	for c := '0'; c <= '9'; c++ {
		count := charCounts[c]
		if float64(count) < float64(expectedPerChar)-tolerance || float64(count) > float64(expectedPerChar)+tolerance {
			t.Logf("Character %c count %d (expected ~%d)", c, count, expectedPerChar)
		}
	}
	for c := 'a'; c <= 'f'; c++ {
		count := charCounts[c]
		if float64(count) < float64(expectedPerChar)-tolerance || float64(count) > float64(expectedPerChar)+tolerance {
			t.Logf("Character %c count %d (expected ~%d)", c, count, expectedPerChar)
		}
	}
}

// --- Length Consistency Tests ---

func TestHashLengthConsistency(t *testing.T) {
	inputs := []string{
		"",
		"a",
		strings.Repeat("a", 100),
		strings.Repeat("a", 10000),
		"日本語",
		"🚀🌍💻",
	}

	for _, input := range inputs {
		hash := HashBody(input)
		if len(hash) != 64 {
			t.Errorf("Input %q produced hash of length %d, expected 64", input[:min(20, len(input))], len(hash))
		}
	}
}

func TestSecretLengthConsistency(t *testing.T) {
	binding := "POST|/api|"
	for i := 0; i < 100; i++ {
		nonce := fmt.Sprintf("%032d", i)
		contextID := fmt.Sprintf("context_%d", i)
		secret := DeriveClientSecret(nonce, contextID, binding)

		if len(secret) != 64 {
			t.Errorf("Secret length %d, expected 64", len(secret))
		}
	}
}

func TestProofLengthConsistency(t *testing.T) {
	secret := strings.Repeat("a", 64)
	binding := "POST|/api|"

	for i := 0; i < 100; i++ {
		timestamp := fmt.Sprintf("%d", i)
		bodyHash := HashBody(fmt.Sprintf(`{"i":%d}`, i))
		proof := BuildProofV21(secret, timestamp, binding, bodyHash)

		if len(proof) != 64 {
			t.Errorf("Proof length %d, expected 64", len(proof))
		}
	}
}

// --- Format Consistency Tests ---

func TestHashFormatLowercaseHex(t *testing.T) {
	for i := 0; i < 100; i++ {
		hash := HashBody(fmt.Sprintf("input_%d", i))

		for j, c := range hash {
			if !((c >= '0' && c <= '9') || (c >= 'a' && c <= 'f')) {
				t.Errorf("Invalid char at position %d: %c", j, c)
			}
		}
	}
}

func TestSecretFormatLowercaseHex(t *testing.T) {
	binding := "POST|/api|"
	for i := 0; i < 100; i++ {
		secret := DeriveClientSecret(fmt.Sprintf("%032d", i), "ctx", binding)

		for j, c := range secret {
			if !((c >= '0' && c <= '9') || (c >= 'a' && c <= 'f')) {
				t.Errorf("Invalid char at position %d: %c", j, c)
			}
		}
	}
}

func TestProofFormatLowercaseHex(t *testing.T) {
	secret := strings.Repeat("a", 64)
	binding := "POST|/api|"

	for i := 0; i < 100; i++ {
		proof := BuildProofV21(secret, fmt.Sprintf("%d", i), binding, HashBody("{}"))

		for j, c := range proof {
			if !((c >= '0' && c <= '9') || (c >= 'a' && c <= 'f')) {
				t.Errorf("Invalid char at position %d: %c", j, c)
			}
		}
	}
}

// --- Sensitivity Tests ---

func TestHashSensitivityToInput(t *testing.T) {
	inputs := []struct {
		a, b string
	}{
		{"hello", "hello "},
		{"hello", "Hello"},
		{"123", "1234"},
		{`{"a":1}`, `{"a":2}`},
		{`{"a":1}`, `{"a": 1}`},
	}

	for _, tc := range inputs {
		hashA := HashBody(tc.a)
		hashB := HashBody(tc.b)

		if hashA == hashB {
			t.Errorf("Different inputs should produce different hashes: %q vs %q", tc.a, tc.b)
		}
	}
}

func TestSecretSensitivityToNonce(t *testing.T) {
	contextID := "test-ctx"
	binding := "POST|/api|"
	nonces := []string{
		strings.Repeat("a", 32),
		strings.Repeat("a", 31) + "b",
		strings.Repeat("b", 32),
	}

	secrets := make(map[string]bool)
	for _, nonce := range nonces {
		secret := DeriveClientSecret(nonce, contextID, binding)
		if secrets[secret] {
			t.Error("Different nonces should produce different secrets")
		}
		secrets[secret] = true
	}
}

func TestSecretSensitivityToContext(t *testing.T) {
	nonce := strings.Repeat("a", 32)
	binding := "POST|/api|"
	contexts := []string{"ctx1", "ctx2", "ctx3"}

	secrets := make(map[string]bool)
	for _, ctx := range contexts {
		secret := DeriveClientSecret(nonce, ctx, binding)
		if secrets[secret] {
			t.Error("Different contexts should produce different secrets")
		}
		secrets[secret] = true
	}
}

// --- Timing Safe Comparison Tests ---

func TestTimingSafeCompareEqualCrypto(t *testing.T) {
	a := strings.Repeat("a", 64)
	b := strings.Repeat("a", 64)

	if !TimingSafeCompare(a, b) {
		t.Error("Equal strings should compare as true")
	}
}

func TestTimingSafeCompareNotEqualCrypto(t *testing.T) {
	a := strings.Repeat("a", 64)
	b := strings.Repeat("b", 64)

	if TimingSafeCompare(a, b) {
		t.Error("Different strings should compare as false")
	}
}

func TestTimingSafeCompareDifferentLengthCrypto(t *testing.T) {
	a := strings.Repeat("a", 64)
	b := strings.Repeat("a", 63)

	if TimingSafeCompare(a, b) {
		t.Error("Different length strings should compare as false")
	}
}

func TestTimingSafeCompareEmptyCrypto(t *testing.T) {
	if !TimingSafeCompare("", "") {
		t.Error("Empty strings should compare as true")
	}
}

func TestTimingSafeCompareSingleDiffCrypto(t *testing.T) {
	a := strings.Repeat("a", 64)

	// Test single character difference at each position
	for i := 0; i < 64; i++ {
		b := a[:i] + "b" + a[i+1:]
		if TimingSafeCompare(a, b) {
			t.Errorf("Single diff at position %d should compare as false", i)
		}
	}
}

// --- Bytes Comparison Tests ---

func TestTimingSafeCompareBytesEqualCrypto(t *testing.T) {
	a := []byte(strings.Repeat("a", 64))
	b := []byte(strings.Repeat("a", 64))

	if !TimingSafeCompareBytes(a, b) {
		t.Error("Equal bytes should compare as true")
	}
}

func TestTimingSafeCompareBytesNotEqualCrypto(t *testing.T) {
	a := []byte(strings.Repeat("a", 64))
	b := []byte(strings.Repeat("b", 64))

	if TimingSafeCompareBytes(a, b) {
		t.Error("Different bytes should compare as false")
	}
}

func TestTimingSafeCompareBytesDifferentLengthCrypto(t *testing.T) {
	a := []byte(strings.Repeat("a", 64))
	b := []byte(strings.Repeat("a", 63))

	if TimingSafeCompareBytes(a, b) {
		t.Error("Different length bytes should compare as false")
	}
}

// --- Additional Crypto Tests ---

func TestProofDependsOnAllInputs(t *testing.T) {
	secret := strings.Repeat("a", 64)
	timestamp := "1234567890"
	binding := "POST|/api|"
	bodyHash := HashBody(`{}`)

	baseProof := BuildProofV21(secret, timestamp, binding, bodyHash)

	// Change secret
	diffSecret := strings.Repeat("b", 64)
	if BuildProofV21(diffSecret, timestamp, binding, bodyHash) == baseProof {
		t.Error("Different secret should produce different proof")
	}

	// Change timestamp
	if BuildProofV21(secret, "9999999999", binding, bodyHash) == baseProof {
		t.Error("Different timestamp should produce different proof")
	}

	// Change binding
	if BuildProofV21(secret, timestamp, "GET|/api|", bodyHash) == baseProof {
		t.Error("Different binding should produce different proof")
	}

	// Change body hash
	diffBodyHash := HashBody(`{"x":1}`)
	if BuildProofV21(secret, timestamp, binding, diffBodyHash) == baseProof {
		t.Error("Different body hash should produce different proof")
	}
}

func TestSecretDependsOnAllInputs(t *testing.T) {
	nonce := strings.Repeat("a", 32)
	contextID := "ctx-123"
	binding := "POST|/api|"

	baseSecret := DeriveClientSecret(nonce, contextID, binding)

	// Change nonce
	diffNonce := strings.Repeat("b", 32)
	if DeriveClientSecret(diffNonce, contextID, binding) == baseSecret {
		t.Error("Different nonce should produce different secret")
	}

	// Change context
	if DeriveClientSecret(nonce, "ctx-456", binding) == baseSecret {
		t.Error("Different context should produce different secret")
	}

	// Change binding
	if DeriveClientSecret(nonce, contextID, "GET|/api|") == baseSecret {
		t.Error("Different binding should produce different secret")
	}
}

// --- Helper ---

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
